core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/batchmanager/BatchManager.java

```diff
@@ -104,111 +103,88 @@ private BatchManager(Builder<RequestT, ResponseT, BatchResponseT> builder) {
         CompletableFuture<ResponseT> response = new CompletableFuture<>();
         try {
             String batchKey = batchKeyMapperFunction.getBatchKey(request);
-            BatchBuffer<RequestT, ResponseT> requestBuffer =
-                requestsAndResponsesMaps.batchBufferByKey(batchKey,
-                                                          () -> scheduleBufferFlush(batchKey, maxBatchOpenInMs.toMillis(),
-                                                                                          scheduledExecutor));
-            requestBuffer.put(request, response);
-
-            int requestsNum = requestsAndResponsesMaps.get(batchKey).requestSize();
-            if (requestsNum > maxBatchItems) {
-                manualFlushBuffer(batchKey);
+            requestsAndResponsesMaps.put(batchKey,
+                                         () -> scheduleBufferFlush(batchKey, maxBatchOpenInMs.toMillis(), scheduledExecutor),
+                                         request,
+                                         response);
+
+            Map<String, BatchingExecutionContext<RequestT, ResponseT>> flushableRequests =
+                requestsAndResponsesMaps.canManualFlush(batchKey, maxBatchItems);
+            if (flushableRequests != null) {
+                manualFlushBuffer(batchKey, flushableRequests);
```

Comment:
How about extracting the log out to a separate method, say `flushBufferIfNeeded(batchKey)`?

```java
    private void flushBufferIfNeed(String batchKey) {
        Map<String, BatchingExecutionContext<RequestT, ResponseT>> flushableRequests =
            requestsAndResponsesMaps.flushableRequests(batchKey, maxBatchItems);
        if (!ColletionUtils.isNullOrEmpty(flushableRequests)) {
            manualFlushBuffer(batchKey, flushableRequests);
        }
    }
```

Reply:
Extracted to a separate `flushBufferIfNeeded()` method